notion_client/lib/api_objects.py

```diff
@@ -0,0 +1,430 @@
+            text=[RichText.from_json(r) for r in d["text"]],
+            children=[Block.from_json(b) for b in children],
+            checked=d.get("checked"),
+            title=d.get("title"),
+        )
+
+
+@dataclass
+class Page(APIObject):
+    parent: PageParent
+    archived: bool
+    properties: Dict[str, "PropertyValue"]
+
+    @classmethod
+    def from_json(cls, d: Dict[str, Any]) -> "Page":
+        return Page(
+            id=d["id"],
+            object="page",
+            archived=d["archived"],
+            created_time=datetime.strptime(d["created_time"], "%Y-%m-%dT%H:%M:%S.%fZ"),
+            last_edited_time=datetime.strptime(
+                d["last_edited_time"], "%Y-%m-%dT%H:%M:%S.%fZ"
+            ),
+            parent=PageParent.from_json(d["parent"]),
+            properties=dict(
+                [(k, property_value_from_json(v)) for (k, v) in d["properties"].items()]
+            ),
+        )
+
+
+@dataclass
+class User(APIObject):
+    type: Optional[UserType]
+    name: Optional[str]
+    avatar_url: Optional[str]
+    person: Optional[Person]
+    bot: Optional[Bot]
+
+    @classmethod
+    def from_json(cls, d: Dict[str, str]) -> "User":
+        user_type = UserType(d.get("type"))
+        return User(
+            id=d["id"],
+            object="user",
+            created_time=datetime.strptime(d["created_time"], "%Y-%m-%dT%H:%M:%S.%fZ"),
+            last_edited_time=datetime.strptime(
+                d["last_edited_time"], "%Y-%m-%dT%H:%M:%S.%fZ"
+            ),
+            type=user_type,
+            name=d.get("name"),
+            avatar_url=d.get("avatar_url"),
+            person=Person(email=d["email"]) if user_type == UserType.person else None,
+            bot=Bot() if user_type == UserType.bot else None,
+        )
+
+
+def property_value_from_json(d: Dict[str, Any]) -> "PropertyValue":
+    property_type = d["type"]
+    property_type_to_class: Dict[str, Any] = {
+        "title": TitlePropertyValue,
+        "rich_text": RichTextPropertyValue,
+        "number": NumberPropertyValue,
+        "select": SelectPropertyValue,
+        "multi_select": MultiselectPropertyValue,
+        "date": DatePropertyValue,
+        "people": PeoplePropertyValue,
+        "file": FilePropertyValue,
+        "checkbox": CheckboxPropertyValue,
+        "url": UrlPropertyValue,
+        "email": EmailPropertyValue,
+        "phone_number": PhoneNumberPropertyValue,
+        "formula": FormulaPropertyValue,
+        "relation": RelationPropertyValue,
+        "rollup": RollupPropertyValue,
+        "created_time": CreatedTimePropertyValue,
+        "created_by": CreatedByPropertyValue,
+        "last_edited_time": LastEditedTimePropertyValue,
+        "last_edited_by": LastEditedByPropertyValue,
+    }
+    print(d)
+    property_value: PropertyValue = property_type_to_class[property_type].from_json(d)
+
+    print(property_value)
+    print()
+    return property_value
+
+
+@dataclass
+class PropertyValue:
```

Comment:
Wouldnt propertyValue be a subclass of the class Property?

The property values that pages have provide all the same information as the properties objects within databases except they have actual value information.

By not doing so, it looks like it may lead to double maintenance.

Example: SelectProperty vs SelectPropertyValue

Reply:
I don't think the default use case makes it a good candidate for inheritance. Although they definitely share some logic that should be pulled out. 

I don't think it's a good candidate because when we receive page JSONs on an API request, we have no context to the additional keys needed to populate the Property Values (e.g. for a number we don't know the format of a Property Value that is received in an database API call). Thus we have alot of nulls in our Page objects and redundancy in keeping the state in both the database objects and page objects locally (for example, what happens if I change the format of a Property Value of a page; in fact how would that actually work in terms of updating the state across all PropertyValues with the one common Property).